CHANGELOG.md

```diff
@@ -10,12 +10,318 @@ New features:
+
+```purescript
+newtype First a = First (Maybe a)
+```
+
+When a parameter appears in the arguments of another type we infer its role from the declaration of that type: `Coercible (First a) (First b)` holds when `Coercible a b` does, even when the newtype constructor is out of scope, because the parameter of `Maybe` is _representational_.
+
+We cannot infer the role of parameters in recursive position, so we default to _phantom_ but usually end up with something else because we keep the most restrictive role a parameter appears at:
+
+```purescript
+data List a = Nil | Cons a (List a)
+```
+
+The parameter appears at _representational_ (under the `Cons` constructor) and _phantom_ (in recursive position) roles so we infer _representational_: `Coercible (List a) (List b)` holds when `Coercible a b` does.
+
+```purescript
+newtype Mu f = In (f (Mu f))
+```
+
+The parameter appears at _representational_ (under the `In` constructor) and _nominal_ (as argument to itself) roles so we infer _nominal_: `Coercible (Mu f) (Mu g)` does not hold, unless `g` is actually `f`.
+
+#### Role annotations
+
+Inferring _nominal_ roles for foreign data types is safe but can be too constraining sometimes. For example this prevents to coerce `Effect Age` to `Effect Int`, even though they actually have the same runtime representation.
+
+The roles of foreign data types can thus be loosened with explicit role annotations, similar to the [RoleAnnotations](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#role-annotations) GHC extension.
+
+Here’s the annotation we added to `Effect` for example:
+
+```purescript
+type role Effect representational
+```
+
+Conversely, we might want to strengthen the roles of parameters with invariants invisible to the type system. Maps are the canonical example of this: the shape of their underlying tree rely on the `Ord` instance of their keys, but the `Ord` instance of a newtype may behave differently than the one of the wrapped type so it would be unsafe to allow coercions between `Map k1 a` and `Map k2 a`, even when `Coercible k1 k2` holds.
+
+In order to forbid such unsafe coercion we added a _nominal_ annotation to the first parameter of `Map`:
+
+```purescript
+type role Map nominal representational
+```
+
+Annotated roles are compared against the roles inferred by the compiler so it is not possible to compromise safety by ascribing too permissive roles, except for foreign types.
+
+### Other changes
+
+#### Breaking
+
+* Add compiler support for `Coercible` based `Newtype` (#3975, @fsoikin)
+
+We added a `Coercible` superclass to `Data.Newtype.Newtype` in order to implement `unwrap`, `wrap` and most newtype combinators with `coerce`. This is a breaking change for non derived instances because the `Newtype` class has no members anymore and can now only be implemented for representationally equal types (those satisfying the new superclass constraint).
```

Comment:
Could we add a brief example of how to do this in a code snippet? For example, you'd no longer implement `wrap` and `unwrap` for the `Additive` newtype:

```diff
  newtype Additive a = Additive a

+ instance newtypeAdditive :: Newtype (Additive a) a
- instance newtypeAdditive :: Newtype (Additive a) a where
-   wrap = Additive
-   unwrap (Additive a) = a
```

Also, it's not clear to me what happens in the case of derived instances right now, like:

```purs
newtype Additive a = Additive a

derive instance newtypeAdditive :: Newtype (Additive a) _
```

Is the derived instance identical to just writing `instance`? Does it warn?

Reply:
I added the example you suggested!

> Also, it's not clear to me what happens in the case of derived instances right now

Derived instances are unaffected. I tried to clarify this by adding a note about that in the changelog. Is that enough?